### PR TITLE
[Kernel] Add Split-KV Support to Unified Triton Attention Kernel

### PR DESCRIPTION
**Pull request by jvlunteren**
In this PR, we introduce performance enhancements to the triton unified attention kernel #16828 by adding a split-KV variant that also parallelizes across the sequence (context) dimension. This approach provides a clear advantage over the current upstream implementation in scenarios involving small batch sizes and long sequences.

This initial version utilizes a simple heuristic to dynamically select between the original and the split-KV kernel versions: If the batch does not contain any prefill requests and the grid size required to launch the current unified attention kernel remains below a certain threshold (indicating limited parallelism), then the new kernel is used. Otherwise, the current unified attention kernel is selected. This approach may be upgraded to a more sophisticated one in the future.

## Performance
The following results were obtained for `meta-llama/Llama-3.1-8B-Instruct` on an NVIDIA H100 GPU, by running

<pre>$ VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1 VLLM_USE_V1=1 python benchmarks/benchmark_latency.py \
    --model meta-llama/Llama-3.1-8B-Instruct \
    --input-len 500 --output-len &ltoutput-length&gt \
    --batch-size &ltbatch-size&gt </pre>

for
1. the V1 `FlashAttention` backend (`VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1` is not used in the above command),
2.  the current triton unified attention kernel, and
3. the new triton unified attention kernel (this PR).

Results for a batch size 1 are shown in the following graph. The output length (in tokens) was varied in these experiments across the following values: 10, 100, 200, 400, 800, 1600, 3200, 6400, and 12800. The number of warmup iterations and measurement iterations were left at the default values of 10 and 30 respectively.

benchmark_latency_batch-size=1

As shown in the graph above, this PR improves the performance of the triton unified attention kernel by more than a factor of two for batch size 1 and an output length of 12800 tokens, also narrowing the performance gap with the flash attention kernel. The performance gains are expected to be even more significant for longer output lengths, compared to the current upstream implementation.

For larger batch sizes, the heuristic selects the current unified kernel, thereby preserving existing performance. Therefore, it remains to be demonstrated that for smaller batch sizes and shorter sequences, for which the new unified kernel is selected but not expected to excel, performance slightly improves or, at the very least, does not degrade compared to the current implementation.

This is now demonstrated through the following experiment involving `mistralai/Mistral-Small-24B-Instruct-2501` on an H100, which was also used in #16828 .

A server is launched using the following command:
<pre>$ VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1 VLLM_USE_V1=1 vllm serve \
        mistralai/Mistral-Small-24B-Instruct-2501 \
        --disable-log-requests \
        --no-enable-prefix-caching
</pre>

Next, the following script is executed:
<pre>MODEL=mistralai/Mistral-Small-24B-Instruct-2501
REQUEST_RATES=(1 5 7 9)
TOTAL_SECONDS=120

for REQUEST_RATE in "${REQUEST_RATES[@]}";
do
    NUM_PROMPTS=$(($TOTAL_SECONDS * $REQUEST_RATE))

    echo ""
    echo "===== RUNNING $MODEL FOR $NUM_PROMPTS PROMPTS WITH $REQUEST_RATE QPS ====="
    echo ""

    python3 benchmark_serving.py \
        --model $MODEL \
        --dataset-name random \
        --request-rate $REQUEST_RATE \
        --random-input-len 1000 \
        --random-output-len 100 \
        --num-prompts $NUM_PROMPTS \
        --ignore-eos --seed $REQUEST_RATE
done</pre>

The measured mean TTFT and ITL values are presented in the following graphs:

ttft

itl

The graphs confirm that performance remains stable and even improves slightly for shorter sequences, which are not expected to benefit significantly from this PR.

## Correctness

V1 `FlashAttention`:
<pre>VLLM_USE_V1=1 lm_eval --model vllm --model_args pretrained=meta-llama/Llama-3.1-8B-Instruct --tasks gsm8k --num_fewshot 5 --batch_size auto --limit 500

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.794|±  |0.0181|
|     |       |strict-match    |     5|exact_match|↑  |0.772|±  |0.0188|</pre>

This PR:
<pre>VLLM_USE_V1=1 VLLM_ATTENTION_BACKEND=TRITON_ATTN_VLLM_V1 lm_eval --model vllm --model_args pretrained=meta-llama/Llama-3.1-8B-Instruct --tasks gsm8k --num_fewshot 5 --batch_size auto --limit 500

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.798|±  |0.0180|
|     |       |strict-match    |     5|exact_match|↑  |0.782|±  |0.0185|</pre>

## How is this performance improvement achieved?

The current triton unified attention kernel uses a 2D launch grid. To scale this further, we introduced a 3D grid for parallelizing also in the sequence (context) dimension. This is achieved by splitting the attention computation into multiple segments, and using online softmax to compute partial results in each segment. A reduction kernel is then used to combine the partial results, enabling correct normalization across the segment boundaries.

cc @tdoublep

### CODE REVIEW

**Comment by robertgshaw2-redhat** [comment]
really nice work

**Review by SageMoore — APPROVED** [review]
LGTM, @jvlunteren. Very nice result! IIRC the prefix_prefil unit test covers this kernel. Can you just double check that to make sure we have appropriate unit test coverage. Otherwise the code looks reasonable.

File: vllm/attention/ops/triton_unified_attention.py [hunk 1]
```diff
@@ -241,6 +242,324 @@ def kernel_unified_attention_2d(
...
+        query_head_idx * (NUM_SEGMENTS_PER_SEQ * HEAD_SIZE_PADDED) +
+        tl.arange(0, NUM_SEGMENTS_PER_SEQ)[:, None] * HEAD_SIZE_PADDED +
+        tl.arange(0, HEAD_SIZE_PADDED)[None, :])
+    segm_output = tl.load(
+        segm_output_ptr + segm_output_offset,
+        mask=segm_mask[:, None] & dim_mask[None, :],
+        other=0.0,
+    )
+    segm_output *= tl.exp(segm_max - overall_max)[:, None]
+    acc = tl.sum(segm_output, axis=0) / overall_expsum
```
[the remark this answers is not in the document]
Comment (jvlunteren) [reply vllm/attention/ops/triton_unified_attention.py]: done

File: vllm/attention/ops/triton_unified_attention.py [hunk 2]
```diff
@@ -241,6 +242,326 @@ def kernel_unified_attention_2d(
...
+    right = num_seqs
+    while left < right:
+        mid = (left + right) // 2
+        mid_val = tl.load(query_start_len_ptr + mid) // BLOCK_Q + mid
+        if mid_val <= q_block_global_idx:
+            left = mid + 1
+        else:
+            right = mid
+
+    seq_idx = left - 1
```
Comment (tlrmchlsmth) [inline vllm/attention/ops/triton_unified_attention.py]: It'd be nice to factor this out into a helper function (since it appears twice)

That would help the code be more self-documenting as well.

File: vllm/attention/ops/triton_unified_attention.py [hunk 3]
```diff
@@ -241,6 +242,326 @@ def kernel_unified_attention_2d(
...
+        (num_query_heads * NUM_SEGMENTS_PER_SEQ * HEAD_SIZE_PADDED) +
+        query_offset_1[:, None] * (NUM_SEGMENTS_PER_SEQ * HEAD_SIZE_PADDED) +
+        segm_idx * HEAD_SIZE_PADDED + tl.arange(0, HEAD_SIZE_PADDED)[None, :])
+    tl.store(
+        segm_output_ptr + segm_output_offset,
+        acc,
+        mask=dim_mask[None, :] & query_mask_0[:, None] & query_mask_1[:, None],
+    )
+    segm_offset = (query_offset_0 * (num_query_heads * NUM_SEGMENTS_PER_SEQ) +
+                   query_offset_1 * NUM_SEGMENTS_PER_SEQ + segm_idx)
```
Comment (tlrmchlsmth) [inline vllm/attention/ops/triton_unified_attention.py]: Not 100% sure but I think the `segm_output_offset` and `segm_offset` calculations could be in int32 with risk of overflow. Could you ensure this happens in int64? 
Changing `query_offset_0[:, None]` to `query_offset_0[:, None].to(tl.int64)` should be sufficient.

File: vllm/attention/ops/triton_unified_attention.py [hunk 4]
```diff
@@ -241,6 +242,326 @@ def kernel_unified_attention_2d(
...
+    segm_expsum = segm_expsum * tl.exp(segm_max - overall_max)
+    overall_expsum = tl.sum(segm_expsum)
+
+    # load, rescale, and add segment attention outputs
+    segm_output_offset = (
+        token_idx *
+        (num_query_heads * NUM_SEGMENTS_PER_SEQ * HEAD_SIZE_PADDED) +
+        query_head_idx * (NUM_SEGMENTS_PER_SEQ * HEAD_SIZE_PADDED) +
+        tl.arange(0, NUM_SEGMENTS_PER_SEQ)[:, None] * HEAD_SIZE_PADDED +
+        tl.arange(0, HEAD_SIZE_PADDED)[None, :])
```
Comment (tlrmchlsmth) [inline vllm/attention/ops/triton_unified_attention.py]: Similar comment as above - I think `segm_offset` and `segm_output_offset` have risk of i32 overflow

**Review by tlrmchlsmth — COMMENTED** [review]
Nice work! Left some comments on the stride calculations that should be addressed before landing

File: vllm/attention/ops/triton_unified_attention.py (hunk 2 above)
Comment (jvlunteren) [reply vllm/attention/ops/triton_unified_attention.py]: Done

File: vllm/attention/ops/triton_unified_attention.py (hunk 3 above)
Comment (jvlunteren) [reply vllm/attention/ops/triton_unified_attention.py]: Done

File: vllm/attention/ops/triton_unified_attention.py (hunk 4 above)
Comment (jvlunteren) [reply vllm/attention/ops/triton_unified_attention.py]: Done

**Review by tlrmchlsmth — APPROVED** [review]
LGTM, thanks and nice work!
